power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +284,598 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+                NodeState<sym>{.u = u_pu_controlled_node<TransformerType>(regulator, state, math_output),
+                               .i = i_pu_controlled_node<TransformerType>(regulator, state, math_output)};
+
+            auto const cmp = node_state <=> param;
+            auto new_tap_pos = [&transformer, &cmp] {
+                if (cmp > 0) { // NOLINT(modernize-use-nullptr)
+                    return one_step_voltage_down(transformer);
+                }
+                if (cmp < 0) { // NOLINT(modernize-use-nullptr)
+                    return one_step_voltage_up(transformer);
+                }
+                return transformer.tap_pos();
+            }();
+
+            if (new_tap_pos != transformer.tap_pos()) {
+                add_tap_pos_update(new_tap_pos, transformer, update_data);
+                tap_changed = true;
+            }
+        });
+
+        return tap_changed;
+    }
+
+    void update_state(UpdateBuffer const& update_data) const {
+        static_assert(sizeof...(TransformerTypes) == std::tuple_size_v<UpdateBuffer>);
+
+        ConstDataset update_dataset;
+        auto const update_component = [this, &update_data, &update_dataset]<transformer_c TransformerType>() {
+            auto const& component_update = get<TransformerType>(update_data);
+            if (!component_update.empty()) {
+                update_dataset.emplace(TransformerType::name, this->get_data_ptr<TransformerType>(update_data));
+            }
+        };
+        (update_component.template operator()<TransformerTypes>(), ...);
+
+        if (!update_dataset.empty()) {
+            update_(update_dataset);
+        }
+    }
+
+    auto initialize(std::vector<std::vector<RegulatedTransformer>> const& regulator_order) const {
+        using namespace std::string_literals;
+
+        constexpr auto max_voltage_pos = [](transformer_c auto const& transformer) -> IntS {
+            // max voltage at control side => min tap pos
+            return transformer.tap_min();
+        };
+        constexpr auto min_voltage_pos = [](transformer_c auto const& transformer) -> IntS {
+            // min voltage at control side => max tap pos
+            return transformer.tap_max();
+        };
+
+        switch (strategy_) {
+        case OptimizerStrategy::any:
+            break;
+        case OptimizerStrategy::global_maximum:
+            [[fallthrough]];
+        case OptimizerStrategy::local_maximum:
+            adjust_voltage(max_voltage_pos, regulator_order);
+            break;
+        case OptimizerStrategy::global_minimum:
+            [[fallthrough]];
+        case OptimizerStrategy::local_minimum:
+            adjust_voltage(min_voltage_pos, regulator_order);
+            break;
+        default:
+            throw MissingCaseForEnumError{"TapPositionOptimizer::initialize"s, strategy_};
+        }
+    }
+
+    void step_all_back(std::vector<std::vector<RegulatedTransformer>> const& regulator_order) const {
```

Comment:
I do not quite understand the new function name? 
And honestly I do not see why this function call is necessary in the first place.

(if this call is indeed necessary) this function could be re-configured for the `initialize` function?

Reply:
maybe better: `reinitialize` or `initialize_refinement`?

Reply:
`exploit_neighborhood`